### ISSUE TITLE
adding wait-for and netcat to opsdroid dockerfile

### ISSUE THREAD

**Issue by bparbhu** [opsdroid/opsdroid#1881]
Hi all,

So I'm still working on getting an opsdroid-rasa swarm up and running and one of the things that I think might help would be to incorporate a tool of some kind to help opsdroid to controlled deployments.  Many of the examples I've seen with rasa and opsdroid involve someone writing either two shell scripts to spin up their rasa container first and then launch their opsdroid container of which doesn't create a problem where opsdroid can't detect the rasa instance. 

Though, in doing a docker-compose file-based launch of rasa with opsdroid you would need some way to control the opsdroid deployment so that it waits for "no pun intended" the rasa container to be up and running with healthy service. In future versions of docker like 3.9 they use a combination of depends_on with condition and specify the container have a healthy service of some kind. In most of docker version 3 they removed this functionality leading you to rely on one of these tools to do it for you. 

So that's why I went with wait-for https://github.com/eficode/wait-for 

So far it's the most up-to-date tool used for alpine based docker containers like opsdroid. 

Right now I'm trying to integrate this through overriding the entrypoint and the command in my docker-compose but for some reason I don't think it's working too well. 

What I think might be a good solution is to incorporate something like this for opsdroid in the docker image so that when someone does need a separate rasa instance that needs to be up and running before their opsdroid launch they have it up and running and they know opsdroid will detect it without error. 

Also this could apply to someone who is also spinning up a database at the same time or spinning up other pieces of hardware that work well with opsdroid so not just rasa but maybe an ngrok container for instance. 

Also happy to help and maybe learn a bit from this in how the opsdroid  is really built. 

Thanks again and always appreciative of the work everyone does here.

-Brian

**Comment by jacobtomlinson** [comment]
I don't think this should be a deployment level thing. You should be able to start Opsdroid and Rasa in any order without things breaking. It sounds like we should add some retries and a configurable timeout to the Rasa code.

We should definitely add this in the training.

https://github.com/opsdroid/opsdroid/blob/a9ab3401202580fb396fdcbb38c85c42583ab0d2/opsdroid/parsers/rasanlu.py#L153

But also in situations where training is skipped what would we expect to happen? If Opsdroid starts before Rasa should we check connectivity and wait before accepting any messages? Or should we start opsdroid up and start processing events but if messages come in before Rasa is ready should we queue those messages until it is ready? I'm curious on what you would expect to happen.

**Comment by bparbhu** [comment]
Hi @jacobtomlinson ,

Thanks again for the feedback, much appreciated! Very cool, if you think this is not a docker-based issue and this can be fixed by just adding some retries and a timeout then I'm down with it. I think I went with a docker-based solution because I wasn't sure if this was something that needed to be handled in the parser codebase for rasa and most likely someone would just wait until the rasa container that would spin up would have healthy service before opsdroid starts. 

When you don't need to train a model one thing I'm assuming is that the person who spins up a rasa instance is already serving up a model on their rasa container. That model gets fed to a certain temporary directory that gets randomly generated by the rasa container each time you spin up a rasa instance. The output looks like this from your rasa container.

rasa        | 2021-12-15 00:54:14 INFO     root  - Starting Rasa server on http://localhost:5005
rasa        | 2021-12-15 00:54:14 DEBUG    rasa.core.utils  - Using the default number of Sanic workers (1).
rasa        | 2021-12-15 00:54:14 INFO     rasa.model  - Loading model ../home/ubuntu/datascience/chatops/opsdroid_deployment/models/slate_faq_model_v5.tar.gz...
rasa        | 2021-12-15 00:54:14 DEBUG    rasa.model  - Extracted model to '/tmp/tmpsuokhfgg'.
rasa        | 2021-12-15 00:54:14 DEBUG    urllib3.connectionpool  - Starting new HTTPS connection (1): api.segment.io:443
rasa        | 2021-12-15 00:54:15 DEBUG    urllib3.connectionpool  - https://api.segment.io:443 "POST /v1/track HTTP/1.1" 200 21
rasa        | 2021-12-15 00:54:15 INFO     root  - Enabling coroutine debugging. Loop id 99882000.
rasa        | 2021-12-15 00:54:15 INFO     rasa.model  - Loading model ../home/ubuntu/datascience/chatops/opsdroid_deployment/models/slate_faq_model_v5.tar.gz...
rasa        | 2021-12-15 00:54:15 DEBUG    rasa.model  - Extracted model to '/tmp/tmp6tk9t9yg'.
rasa        | 2021-12-15 00:54:15 DEBUG    rasa.core.tracker_store  - Connected to InMemoryTrackerStore.
rasa        | 2021-12-15 00:54:15 DEBUG    rasa.core.lock_store  - Connected to lock store 'InMemoryLockStore'.
rasa        | 2021-12-15 00:54:16 DEBUG    rasa.model  - Extracted model to '/tmp/tmparckfd4p'.
rasa        | 2021-12-15 00:54:16 DEBUG    rasa.core.nlg.generator  - Instantiated NLG to 'TemplatedNaturalLanguageGenerator'.
rasa        | 2021-12-15 00:54:16 INFO     root  - Rasa server is up and running.


The filepath is usually tmp/tmpxxxxxx/. 
If you're not training a model that's where you should look for one. 

Something I think could be related to this could be this issue https://github.com/opsdroid/opsdroid/issues/1876

In terms of behavior for how messages get processed in opsdroid before rasa starts up, I haven't thought about that yet. Though, I don't want to assume that someone is running a bot and sending messages to an opsdroid instance before their rasa instance starts up, as they actually put in the effort to training a rasa model, but in the case where maybe a container stopped or didn't start up in time usually what I would like to see is if my rasa instance is actually working first and know that isn't before accepting messages. That way I would double check my connection to the rasa container to see if it's there. Once a user specifies a rasa parser they should know if there is at least a connection to that instance before accepting messages. Though, that's how I feel about at least. I haven't gotten to test this out with multiple users yet so I think that's where we would see behavior like that and most likely I can make a better judgement call after seeing how people use my bot. 

Please let me know if you need more information that could help I'll try and see if I can do a pull request using the solution you mentioned.

Thanks again @jacobtomlinson , much appreciated!

-Brian

**Comment by jacobtomlinson** [comment]
Yeah I think in the load method we should have a `setup_parsers` function which does anything that is needed before parsers can be used.

https://github.com/opsdroid/opsdroid/blob/a9ab3401202580fb396fdcbb38c85c42583ab0d2/opsdroid/core.py#L217

In this case if the Rasa parser is enabled it should just check that it can connect and if not wait for it to connect. It is probably best to add a timeout too.

If you feel like taking it on and raising a PR it would be much appreciated.

**Comment by bparbhu** [comment]
Will do, I'll give it a shot @jacobtomlinson !

**Comment by bparbhu** [comment]
Hmm so I have been trying to get this to work but I could use a second set of eyes. It's mainly dealing with the rasanlu parser codebase. The parts I have decided to change deal with how aiohttp sessions do responses. It's there where I substitute for aiohttp_retry versions of responses but ultimately I do get an error around having sessions that aren't closed. Though, its weird in that I specifically close each response session after I finish using them. I didn't see many sessions or any being closed at all using the original aiohttp responses. Does opsdroid allow for sessions to be open and also do you think there's anything I'm obviously doing wrong. The link to the forked repo is here https://github.com/bparbhu/opsdroid. My changes are under the rasanlu parser, the setup config, and I did originally mess around with the Dockerfile but I have since made sure it's back to what it needs to be. I have been doing my testing with Docker-compose and also just building a new docker image after each change.

**Comment by bparbhu** [comment]
Also this is the link to aiohttp_retry https://github.com/inyutin/aiohttp_retry

**Comment by jacobtomlinson** [comment]
@bparbhu feel free to open your changes as a draft Pull Request and we can discuss there. It's much easier for us to see what you've changed that way.

**Comment by bparbhu** [comment]
@jacobtomlinson will do. I don't think these are changes I want to permanently commit but I still need to talk about them.

**Comment by bparbhu** [comment]
Hey @jacobtomlinson Just submitted the draft pull request. Again just need some feedback and some better direction.

**Comment by bparbhu** [comment]
So I did get the rasa instance to work with opsdroid within docker-compose. All it took as just adding the timeout flags where your responses occur and also I had to make an entrypoint script that sleeps opsdroid for 60 seconds and then starts opsdroid with the config I want. Still testing out rasa skills but I think the fix looks promising.
